File: pysrc/Image.cpp
/* -*- c++ -*-
 * Copyright (c) 2012-2023 by the GalSim developers team on GitHub
 * https://github.com/GalSim-developers
 *
 * This file is part of GalSim: The modular galaxy image simulation toolkit.
 * https://github.com/GalSim-developers/GalSim
 *
 * GalSim is free software: redistribution and use in source and binary forms,
 * with or without modification, are permitted provided that the following
 * conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions, and the disclaimer given in the accompanying LICENSE
 *    file.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions, and the disclaimer given in the documentation
 *    and/or other materials provided with the distribution.
 */

#include "PyBind11Helper.h"
#include "Image.h"

// Note that docstrings are now added in galsim/image.py
namespace galsim {

    template <typename T>
    static ImageView<T>* MakeFromArray(
        size_t idata, int step, int stride, const Bounds<int>& bounds)
    {
        T* data = reinterpret_cast<T*>(idata);
        // Note: use ptrdiff_t for these to make sure the product that becomes
        // the maxptr offset or nElements doesn't overflow.
        ptrdiff_t ncol = bounds.getXMax() - bounds.getXMin() + 1;
        ptrdiff_t nrow = bounds.getYMax() - bounds.getYMin() + 1;
        const T* maxptr = data + (ncol-1) * step + (nrow-1) * stride + 1;
        shared_ptr<T> owner;
        return new ImageView<T>(data, maxptr, ncol*nrow, owner, step, stride, bounds);
    }

    template <typename T>
    static void DepixelizeImage(ImageView<T> im, size_t iunit_integrals, const int n)
    {
        const double* unit_integrals = reinterpret_cast<const double*>(iunit_integrals);
        im.depixelizeSelf(unit_integrals, n);
    }

    template <typename T>
    static void WrapImage(py::module& _galsim, const std::string& suffix)
    {
        py::class_<BaseImage<T> >(_galsim, ("BaseImage" + suffix).c_str());

        typedef ImageView<T>* (*Make_func)(size_t, int, int, const Bounds<int>&);
        py::class_<ImageView<T>, BaseImage<T> >(_galsim, ("ImageView" + suffix).c_str())
            .def(py::init((Make_func)&MakeFromArray));

        typedef void (*rfft_func_type)(const BaseImage<T>&, ImageView<std::complex<double> >,
                                       bool, bool);
        typedef void (*irfft_func_type)(const BaseImage<T>&, ImageView<double>, bool, bool);
        typedef void (*cfft_func_type)(const BaseImage<T>&, ImageView<std::complex<double> >,
                                       bool, bool, bool);
        _galsim.def("rfft", rfft_func_type(&rfft));
        _galsim.def("irfft", irfft_func_type(&irfft));
        _galsim.def("cfft", cfft_func_type(&cfft));

        typedef void (*wrap_func_type)(ImageView<T>, const Bounds<int>&, bool, bool);
        _galsim.def("wrapImage", wrap_func_type(&wrapImage));

        typedef void (*invert_func_type)(ImageView<T>);
        _galsim.def("invertImage", invert_func_type(&invertImage));

        typedef void (*depix_func_type)(ImageView<T>, size_t, const int);
        _galsim.def("depixelizeImage", depix_func_type(&DepixelizeImage));
    }

    void pyExportImage(py::module& _galsim)
    {
        WrapImage<uint16_t>(_galsim, "US");
        WrapImage<uint32_t>(_galsim, "UI");
        WrapImage<int16_t>(_galsim, "S");
        WrapImage<int32_t>(_galsim, "I");
        WrapImage<float>(_galsim, "F");
        WrapImage<double>(_galsim, "D");
        WrapImage<std::complex<double> >(_galsim, "CD");
        WrapImage<std::complex<float> >(_galsim, "CF");

        _galsim.def("goodFFTSize", &goodFFTSize);
        _galsim.def("ClearDepixelizeCache", &ClearDepixelizeCache);
    }

} // namespace galsim

File: galsim/deprecated/image.py
# Copyright (c) 2012-2017 by the GalSim developers team on GitHub
# https://github.com/GalSim-developers
#
# This file is part of GalSim: The modular galaxy image simulation toolkit.
# https://github.com/GalSim-developers/GalSim
#
# GalSim is free software: redistribution and use in source and binary forms,
# with or without modification, are permitted provided that the following
# conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
#    list of conditions, and the disclaimer given in the accompanying LICENSE
#    file.
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions, and the disclaimer given in the documentation
#    and/or other materials provided with the distribution.
#

import galsim
import numpy
from galsim.deprecated import depr

# Don't forget to also remove the MetaImage stuff from image.py when we get rid of this
# round of deprecations.  (It doesn't really work to insert a metaclass after the class
# is already made, so I couldn't put the MetaImage stuff here.)

def ImageViewS(*args, **kwargs):
    """A deprecated alias for galsim.Image(..., dtype=numpy.int16)
    """
    depr('ImageView[type] or ImageViewS', 1.1, 'ImageS')
    kwargs['dtype'] = numpy.int16
    return galsim.Image(*args, **kwargs)

def ImageViewI(*args, **kwargs):
    """A deprecated alias for galsim.Image(..., dtype=numpy.int32)
    """
    depr('ImageView[type] or ImageViewI', 1.1, 'ImageI')
    kwargs['dtype'] = numpy.int32
    return galsim.Image(*args, **kwargs)

def ImageViewF(*args, **kwargs):
    """A deprecated alias for galsim.Image(..., dtype=numpy.float32)
    """
    depr('ImageView[type] or ImageViewF', 1.1, 'ImageF')
    kwargs['dtype'] = numpy.float32
    return galsim.Image(*args, **kwargs)

def ImageViewD(*args, **kwargs):
    """A deprecated alias for galsim.Image(..., dtype=numpy.float64)
    """
    depr('ImageView[type] or ImageViewD', 1.1, 'ImageD')
    kwargs['dtype'] = numpy.float64
    return galsim.Image(*args, **kwargs)

def ConstImageViewS(*args, **kwargs):
    """A deprecated alias for galsim.Image(..., dtype=numpy.int16, make_const=True)
    """
    depr('ConstImageView[type] or ConstImageViewS', 1.1, 'ImageS(..., make_const=True)')
    kwargs['dtype'] = numpy.int16
    kwargs['make_const'] = True
    return galsim.Image(*args, **kwargs)

def ConstImageViewI(*args, **kwargs):
    """A deprecated alias for galsim.Image(..., dtype=numpy.int32, make_const=True)
    """
    depr('ConstImageView[type] or ConstImageViewI', 1.1, 'ImageI(..., make_const=True)')
    kwargs['dtype'] = numpy.int32
    kwargs['make_const'] = True
    return galsim.Image(*args, **kwargs)

def ConstImageViewF(*args, **kwargs):
    """A deprecated alias for galsim.Image(..., dtype=numpy.float32, make_const=True)
    """
    depr('ConstImageView[type] or ConstImageViewF', 1.1, 'ImageF(..., make_const=True)')
    kwargs['dtype'] = numpy.float32
    kwargs['make_const'] = True
    return galsim.Image(*args, **kwargs)

def ConstImageViewD(*args, **kwargs):
    """A deprecated alias for galsim.Image(..., dtype=numpy.float64, make_const=True)
    """
    depr('ConstImageView[type] or ConstImageViewD', 1.1, 'ImageD(..., make_const=True)')
    kwargs['dtype'] = numpy.float64
    kwargs['make_const'] = True
    return galsim.Image(*args, **kwargs)

galsim.ImageViewS = ImageViewS
galsim.ImageViewI = ImageViewI
galsim.ImageViewF = ImageViewF
galsim.ImageViewD = ImageViewD
galsim.ConstImageViewS = ConstImageViewS
galsim.ConstImageViewI = ConstImageViewI
galsim.ConstImageViewF = ConstImageViewF
galsim.ConstImageViewD = ConstImageViewD

galsim.ImageView = {
    numpy.int16 : ImageViewS,
    numpy.int32 : ImageViewI,
    numpy.float32 : ImageViewF,
    numpy.float64 : ImageViewD
}

galsim.ConstImageView = {
    numpy.int16 : ConstImageViewS,
    numpy.int32 : ConstImageViewI,
    numpy.float32 : ConstImageViewF,
    numpy.float64 : ConstImageViewD
}

def at(self, x, y):
    """This method is a synonym for im(x,y).  It is a bit faster than im(x,y), since GalSim
    does not have to parse the different options available for __call__.  (i.e. im(x,y) or
    im(pos) or im(x=x,y=y))
    """
    depr('image.at(x,y)', 1.5, 'image.getValue(x,y)')
    return self.getValue(x,y)

galsim.Image.at = at

def Image_getXMin(self):
    depr('image.getXMin()', 1.5, 'image.xmin')
    return self.xmin

def Image_getXMax(self):
    depr('image.getXMax()', 1.5, 'image.xmax')
    return self.xmax

def Image_getYMin(self):
    depr('image.getYMin()', 1.5, 'image.ymin')
    return self.ymin

def Image_getYMax(self):
    depr('image.getYMax()', 1.5, 'image.ymax')
    return self.ymax

def Image_getBounds(self):
    depr('image.getBounds()', 1.5, 'image.bounds')
    return self.bounds

galsim.Image.getXMin = Image_getXMin
galsim.Image.getXMax = Image_getXMax
galsim.Image.getYMin = Image_getYMin
galsim.Image.getYMax = Image_getYMax
galsim.Image.getBounds = Image_getBounds
